sources/input/scene_loader.c

```diff
@@ -5,35 +5,124 @@
 
-int	scene_open(char *filename)
+#define SCENE_EOF -1
+#define SCENE_EMPTY_LN 1 
+
+void	free_resources(char *line, char **line_fields);
+int		get_line_fields(int scene_fd, char **line, char ***line_fields);
+int		get_uniq_elem(enum e_element e_type, char **line_f, t_scene *scene);
+int		get_primitive(enum e_element e_type, char **line_f, t_scene *scene);
+
+int	scene_load(int scene_fd, t_scene *scene)
 {
-	int	scene_fd;
+	enum e_element	e_type;
+	int				op_code;
+	char			*line;
+	char			**line_fields;
 
-	scene_fd = open(filename, O_WRONLY);
-	if (scene_fd < 0 && errno == EISDIR)
+	while (42)
 	{
-		print_wrong_extension();
-		return (scene_fd);
+		op_code = get_line_fields(scene_fd, &line, &line_fields);
+		if (op_code == SCENE_EOF)
+			break ;
+		if (op_code == SCENE_EMPTY_LN)
+			continue ;
+		e_type = get_element(line_fields[0]);
+		if (e_type == nae || get_uniq_elem(e_type, &line_fields[1], scene) \
+			|| get_primitive(e_type, &line_fields[1], scene))
+		{
+			free_resources(line, line_fields);
+			return (1);
+		}
+		free_resources(line, line_fields);
 	}
-	close(scene_fd);
-	scene_fd = open(filename, O_RDONLY);
-	if (scene_fd >= 0)
-		return (scene_fd);
-	else if (errno == EPERM || errno == EACCES)
-		print_read_error(filename);	
-	else if (errno == ENOENT)
-		print_no_file(filename);
-	else
-		print_error(filename);
-	return (scene_fd);
+	return (0);
 }
+
+int	get_line_fields(int scene_fd, char **line, char ***line_fields)
+{
+	int	line_size;
+
+	line_size = 0;
+	*line = get_next_line(scene_fd);
+	if (!(*line))
+		return (SCENE_EOF);
+	if (**line == '\n')
+	{
+		free(*line);
+		return (SCENE_EMPTY_LN);
+	}
+	line_size = ft_strlen(*line);
+	if (line_size > 0 && line[0][line_size - 1] == '\n')
+		line[0][line_size - 1] = '\0';
+	*line_fields = ft_split(*line, ' ');
+	if (**line_fields == NULL || *line_fields[0][0] == '\0')
+	{
+		free_resources(*line, *line_fields);
+		return (SCENE_EMPTY_LN);
+	}
+	return (0);
+}
+
+void	free_resources(char *line, char **line_fields)
+{
+	free(line);
+	ft_free_array(line_fields);
+}
+
+int	get_uniq_elem(enum e_element e_type, char **line_f, t_scene *scene)
+{
+	if (e_type == camera && scene->camera != NULL)
+		return (1);
+	if (e_type == light && scene->light != NULL)
+		return (1);
+	if (e_type == ambient_light && scene->a_light != NULL)
+		return (1);
+	if (e_type == camera)
+	{
+		if (init_camera(line_f, &scene->camera))
+			return (1);
+	}
+	else if (e_type == light)
+	{
+		if (init_light(line_f, &scene->light))
+			return (1);
+	}
+	else if (e_type == ambient_light)
+	{
+		if (init_ambient_light(line_f, &scene->a_light))
+			return (1);
+	}
+	return (0);
+}
+
+int	get_primitive(enum e_element e_type, char **line_f, t_scene *scene)
+{
+	t_elist	*new_elem;
+
+	new_elem = NULL;
+	if (e_type != ambient_light && e_type != camera && e_type != light)
+	{
+		new_elem = elist_new(e_type, line_f);
+		if (!new_elem)
+			return (1);
+		elist_addback(&scene->elements, new_elem);
+	}
+	return (0);
+}
+
```

Comment:
Leftover comments in this file

Reply:
yeah, my bad. I was using that code to debug errors and undefined behaviors and I forgot about it ("because I might just need it in the future" lol)